src/Microsoft.TestPlatform.AdapterUtilities/ManagedNameUtilities/ManagedNameHelper.Reflection.cs

```diff
@@ -266,14 +268,169 @@ private static void AppendTypeString(StringBuilder b, Type type, bool closedType
             }
         }
 
-        private static void AppendNestedTypeName(StringBuilder b, Type type)
+        private static void AppendNamespace(StringBuilder b, string ns)
+        {
+            int start = 0;
+            bool shouldEscape = false;
+
+            for (int i = 0; i <= ns.Length; i++)
+            {
+                if (i == ns.Length || ns[i] == '.')
+                {
+                    if (start != 0)
+                    {
+                        b.Append('.');
+                    }
+
+                    var part = ns.Substring(start, i - start);
+                    if (shouldEscape)
+                    {
+                        NormalizeString(b, part);
+                        shouldEscape = false;
+                    }
+                    else
+                    {
+                        b.Append(part);
+                    }
+
+                    start = i + 1;
+                    continue;
+                }
+
+                shouldEscape = shouldEscape || NeedsEscaping(ns[i], i - start);
+            }
+        }
+
+        private static void AppendMethodString(StringBuilder methodBuilder, string name, int methodArity)
         {
+            var arityStart = name.LastIndexOf('`');
+            var arity = 0;
+            if (arityStart > 0)
+            {
+                arityStart++;
+                var arityString = name.Substring(arityStart, name.Length - arityStart);
+                if (int.TryParse(arityString, out arity))
+                {
+                    if (arity == methodArity)
+                    {
+                        name = name.Substring(0, arityStart - 1);
+                    }
+                }
+            }
+
+            if (IsNormalized(name))
+            {
+                methodBuilder.Append(name);
+            }
+            else
+            {
+                NormalizeString(methodBuilder, name);
+            }
+
+            if (arity > 0 && methodArity == arity)
+            {
+                methodBuilder.Append($"`{arity}");
+            }
+        }
+
+        private static void NormalizeString(StringBuilder b, string name)
```

Comment:
Some documentation explaining the business logic of this method would be nice since it doesn't do trivial work.

Unit tests would be nice too.